### ISSUE TITLE
Poor efficiency while running many worker instance nodes

### ISSUE THREAD

**Issue by dejunzhang** [yahoo/CaffeOnSpark#35]
Hi,

I have got a quick result when just start one worker instance node, but the processing became very slow if i change the SPARK_WORKER_INSTANCES to 2.(I have built a 3 nodes cluster)

for example, the cost time for minist change from 45s (1 node) to 3m47s(2 nodes) and for cifar-10 change from 1'20s (1 node) to 7m10s(2 nodes)

I have export raw data to HDFS using the following cmd and set 3 replications.
hadoop fs -put -f ${CAFFE_ON_SPARK}/data/cifar10_*_lmdb hdfs:/projects/machine_learning/image_dataset/

The following is my trainning command for cifar:

export SPARK_WORKER_INSTANCES=2
export DEVICES=1
hadoop fs -rm -r -f hdfs:///cifar10_features_result
hadoop fs -rm hdfs:///cifar10*
spark-submit --master yarn --deploy-mode cluster --num-executors ${SPARK_WORKER_INSTANCES} \
--files ./data/cifar10_quick_solver_hdfs.prototxt,./data/cifar10_quick_train_test_hdfs.prototxt,./data/mean.binaryproto \
--conf spark.driver.extraLibraryPath="${LD_LIBRARY_PATH}" \
--conf spark.executorEnv.LD_LIBRARY_PATH="${LD_LIBRARY_PATH}" \
--class com.yahoo.ml.caffe.CaffeOnSpark ${CAFFE_ON_SPARK}/caffe-grid/target/caffe-grid-0.1-SNAPSHOT-jar-with-dependencies.jar \
-train -features accuracy,loss -label label -conf cifar10_quick_solver_hdfs.prototxt \
-devices ${DEVICES} \
-connection ethernet -model hdfs:///cifar10.model.h5 -output hdfs:///cifar10_features_result

Here is some log from worker node when running multi nodes.

16/03/22 11:05:30 INFO storage.BlockManager: Found block rdd_0_0 locally
I0322 11:05:31.616941 24799 solver.cpp:237] Iteration 1000, loss = 0.935601
I0322 11:05:31.617007 24799 solver.cpp:253] Train net output #0: loss = 0.935601 (* 1 = 0.935601 loss)
I0322 11:05:31.657905 24799 sgd_solver.cpp:106] Iteration 1000, lr = 0.001
I0322 11:05:41.364012 24799 solver.cpp:237] Iteration 1100, loss = 0.955443
I0322 11:05:41.364084 24799 solver.cpp:253] Train net output #0: loss = 0.955443 (* 1 = 0.955443 loss)
I0322 11:05:41.411869 24799 sgd_solver.cpp:106] Iteration 1100, lr = 0.001
I0322 11:05:51.151032 24799 solver.cpp:237] Iteration 1200, loss = 0.813748
I0322 11:05:51.151115 24799 solver.cpp:253] Train net output #0: loss = 0.813748 (* 1 = 0.813748 loss)
I0322 11:05:51.197666 24799 sgd_solver.cpp:106] Iteration 1200, lr = 0.001

Here is some log from worker node when running single node.

16/03/22 14:14:30 INFO storage.BlockManager: Found block rdd_0_0 locally
I0322 14:14:30.212993 9250 solver.cpp:237] Iteration 1000, loss = 0.875192
I0322 14:14:30.213027 9250 solver.cpp:253] Train net output #0: loss = 0.875192 (* 1 = 0.875192 loss)
I0322 14:14:30.213042 9250 sgd_solver.cpp:106] Iteration 1000, lr = 0.001
I0322 14:14:31.229243 9250 solver.cpp:237] Iteration 1100, loss = 1.05669
I0322 14:14:31.229284 9250 solver.cpp:253] Train net output #0: loss = 1.05669 (* 1 = 1.05669 loss)
I0322 14:14:31.229298 9250 sgd_solver.cpp:106] Iteration 1100, lr = 0.001
I0322 14:14:32.238924 9250 solver.cpp:237] Iteration 1200, loss = 0.965516
I0322 14:14:32.238962 9250 solver.cpp:253] Train net output #0: loss = 0.965516 (* 1 = 0.965516 loss)
I0322 14:14:32.238976 9250 sgd_solver.cpp:106] Iteration 1200, lr = 0.001
I0322 14:14:33.250813 9250 solver.cpp:237] Iteration 1300, loss = 0.806154
I0322 14:14:33.250855 9250 solver.cpp:253] Train net output #0: loss = 0.806154 (* 1 = 0.806154 loss)
I0322 14:14:33.250871 9250 sgd_solver.cpp:106] Iteration 1300, lr = 0.001
I0322 14:14:34.266059 9250 solver.cpp:237] Iteration 1400, loss = 0.807488
I0322 14:14:34.266103 9250 solver.cpp:253] Train net output #0: loss = 0.807488 (* 1 = 0.807488 loss)
I0322 14:14:34.266119 9250 sgd_solver.cpp:106] Iteration 1400, lr = 0.001

As you can see, the different between that is every iteration cost time(9s vs 1s). I know there are extra communication between differents nodes, but why so slow?

Looking forward to your response, thx!

**Comment by anfeng** [comment]
@dejunzhang What's is the network bandwidth b/w your nodes? It's very likely your network is congested. It will be interesting to get some network usage info.

May I assume that both runs use identical batch size? Batch size is for per worker instance. You should adjust batch size by dividing it by $(SPARK_WORKER_INSTANCES).

**Comment by junshi15** [comment]
Several factors could contribute to the slowness:

1) weakest node in the cluster. Training is done synchronously, the slowest node brings down the over speed.

2) network bandwidth. Two drawbacks with ethernet connections: (a) GPU memory needs to be copied to host memory before transmission. The copying can be slow. (b) Ethernet is usually slow. We use Infiniband and RDMA for fast memory and network access.

3) as @anfeng mentioned, your effective batch size doubles when you use two nodes since the batch size in the prototxt file is per device.

**Comment by dejunzhang** [comment]
@anfeng the slave nodes are connected with a 1Gb/s switch. Yes, I didn't modify the batch size just now, and they are the same as in single node. I will have a try.  thanks.
@junshi15 the slave nodes have the same hardware and environment. Now we only have ethernet connections between nodes.

**Comment by mriduljain** [comment]
@dejunzhang Let us know if you see differences with above config changes etc, whenever

**Comment by dejunzhang** [comment]
@mriduljain the network usage during the trainning is about 90 ~ 300Mb/s and the time cost is slight higher than the single node but with lower loss. Thank you.
